sdks/python/apache_beam/io/datastore/v1/query_splitter.py

```diff
@@ -0,0 +1,256 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Implements a Cloud Datastore query splitter."""
+
+from apache_beam.io.datastore.v1 import helper
+from google.datastore.v1 import datastore_pb2
+from google.datastore.v1 import query_pb2
+from google.datastore.v1.query_pb2 import PropertyFilter
+from google.datastore.v1.query_pb2 import CompositeFilter
+import googledatastore
+
+
+__all__ = [
+    'get_splits',
+]
+
+SCATTER_PROPERTY_NAME = '__scatter__'
+KEY_PROPERTY_NAME = '__key__'
+# The number of keys to sample for each split.
+KEYS_PER_SPLIT = 32
+
+UNSUPPORTED_OPERATORS = [PropertyFilter.LESS_THAN,
+                         PropertyFilter.LESS_THAN_OR_EQUAL,
+                         PropertyFilter.GREATER_THAN,
+                         PropertyFilter.GREATER_THAN_OR_EQUAL]
+
+
+def get_splits(datastore, query, num_splits, partition=None):
+  """Returns a list of sharded queries for the given Cloud Datastore query.
+
+  This will create up to the desired number of splits, however it may return
+  less splits if the desired number of splits is unavailable. This will happen
+  if the number of split points provided by the underlying Datastore is less
+  than the desired number, which will occur if the number of results for the
+  query is too small.
+
+  This implementation of the QuerySplitter uses the __scatter__ property to
+  gather random split points for a query.
+
+  Note: This implementation is derived from the java query splitter in
+  https://github.com/GoogleCloudPlatform/google-cloud-datastore/blob/master/java/datastore/src/main/java/com/google/datastore/v1/client/QuerySplitterImpl.java
+
+  Args:
+    datastore: the datastore client.
+    query: the query to split.
+    num_splits: the desired number of splits.
+    partition: the partition the query is running in.
+  """
+
+  _validate_query(query)
+  _validate_split_size(num_splits)
```

Comment:
You can inline validation. Especially validate_split_size, is just a single check, it could be done here or in a single validation function.